File: colorful.py
'''
HSPT 2020 - colorful.py
by Jacob Magnuson
(cf\\barakraganosungam)

FYI: The language of this solution departs from the language of the problem
   statement, instead speaking in the terminology of "graph theory". We can
   think of the neighborhood in the problem statement as a network "graph", with
   houses as nodes and paths/edges between them.
'''

# Iterate over each testcase
for tc in range(int(input())):
    N, M, K = map(int, input().split())
    # col is a string of letters A-Z converted into an array with values [0, 25]
    col = [ord(c) - ord('A') for c in input()]
    # adj[v] will be a list of all of the neighbors of vertex v in the graph
    adj = [[] for i in range(N)]
    for m in range(M):
        u, v = (int(x) - 1 for x in input().split())
        adj[u].append(v), adj[v].append(u)
    '''
    We will precompute the answer to all possible queries, by assigning every
       vertex the ID of the closest house of each color.
    We will solve each color independently. To solve a color C, we will spread out
       through the graph in a flood-fill fashion, and set answers as we go.
    We start by making all vertices with color C their own answer. Then, all
       vertices 1 away from any of these vertices are assigned an answer, and so
       forth, in a breadth-first fashion through the graph.
    '''
    near = [[-1] * N for c in range(26)]
    dist = [[2 ** 42] * N for c in range(26)]
    for c in range(26):
        q = []
        # This loop initializes our queue with, and sets the answer for,
        #    all vertices in the graph with color c.
        for v in range(N):
            if(col[v] == c):
                near[c][v] = v + 1
                dist[c][v] = 0
                q.append(v)
        # our BFS/Floodfill loop
        while len(q) != 0:
            nq = []
            # For each item in our queue, we look at all of its neighbors
            for v in q:
                for w in adj[v]:
                    if near[c][w] != -1: continue # Only consider unvisited vertices
                    # Set this vertex's answer to our own. Our ordering ensures
                    #    that this answer will be minimized. Then, add it to the queue.
                    near[c][w] = near[c][v]
                    nq.append(w)
            # Next time, we'll iterate over our queue in order of the base vertex
            #    that they point back to. This means any answers we set will be
            #    minimized (as they will all be equidistant from a color C vertex)
            q = [p for p in sorted(nq, key=lambda z: near[c][z])]
    # With all of our answers saved, we can respond to the queries as they come in now
    print("Neighborhood #%d:" % (tc + 1))
    for k in range(K):
        dat = input().split()
        v, c = int(dat[0]) - 1, ord(dat[1]) - ord('A')
        print(near[c][v])
    print()

File: colorful.cpp
#include <cstdio>
#include <cstdlib>
#include <vector>
#include <queue>

#define MAX_N 100000

using namespace std;

int main(){
    int T, N, M, K;
    char labels[MAX_N+1];

    // input
    scanf("%d\n", &T);

    for(int t = 1; t <= T; t++){

        // more input
        scanf("%d %d %d", &N, &M, &K);

        // adj holds, for each house, its neighbors.
        // dist is distance from a particular color.
        // who is the house of a particular color closest to a house.
        vector<vector<int> > adj;
        vector<vector<int> > dist;
        vector<vector<int> > who;

        for(int n=0;n<N;n++){
            adj.push_back(vector<int>());
        }
        for(int c=0;c<26;c++){
            dist.push_back(vector<int>());
            who.push_back(vector<int>());
            for(int n=0;n<N;n++){
                // note that 999,999,999 is used as infinity.
                dist[c].push_back(999999999);
                who[c].push_back(-2);
            }
        }

        scanf("%s", labels);
        for(int m=0;m<M;m++){
            int a, b;
            scanf("%d %d", &a, &b);
            a--; b--;
            adj[a].push_back(b);
            adj[b].push_back(a);
        }

        // set the distances for all starts
        for(int n = 0; n < N; n++){
            dist[labels[n]-'A'][n] = 0;
        }


        // the plan:
        // get the distance from a color for each color.
        // this gives you a 26 by N matrix of distances.

        // while you BFS, you can also record which house is closest.
        for(int ch = 0; ch < 26; ch++){
            queue<int> queue;

            for(int n=0;n<N;n++){
                // put it onto the queue
                // but only if the character matches
                if(labels[n] == ch+'A'){
                    queue.push(n);
                    who[ch][n] = n;
                }
            }

            while(queue.size() > 0){
                int u = queue.front();
                queue.pop();
                for(int i = 0; i < adj[u].size(); i++){
                    int v = adj[u][i];
                    // visited?
                    if(dist[ch][v] == 999999999){
                        dist[ch][v] = dist[ch][u]+1;
                        who[ch][v] = who[ch][u];
                        queue.push(v);
                    }
                    // of a lower index?
                    else if(
                            dist[ch][u] + 1 == dist[ch][v] &&
                            who[ch][u] < who[ch][v]){
                        who[ch][v] = who[ch][u];
                    }
                }
            }
        }
        printf("Neighborhood #%d:\n", t);
        for(int k=0;k<K;k++){
            int q;
            char ch;

            scanf("%d %c", &q, &ch);

            printf("%d\n", 1+who[ch-'A'][q-1]);
        }
        printf("\n");
    }
}
